contracts/modules/Wallet/VestingEscrowWallet.sol

```diff
@@ -0,0 +1,525 @@
+     * @param _index index of the template
+     * @param _startTime vesting start time
+     */
+    function addScheduleFromTemplate(address _beneficiary, uint256 _index, uint256 _startTime) public withPerm(ADMIN) {
+        require(_index < templates.length, "Template not found");
+        Template storage template = templates[_index];
+        addSchedule(_beneficiary, template.numberOfTokens, template.duration, template.frequency, _startTime);
+    }
+
+    /**
+     * @notice Edits vesting schedules for each of beneficiary
+     * @param _beneficiary beneficiary's addresses
+     * @param _index index of schedule
+     * @param _numberOfTokens number of tokens
+     * @param _duration vesting duration
+     * @param _frequency vesting frequency
+     * @param _startTime vesting start time
+     */
+    function editSchedule(
+        address _beneficiary,
+        uint256 _index,
+        uint256 _numberOfTokens,
+        uint256 _duration,
+        uint256 _frequency,
+        uint256 _startTime
+    )
+        public
+        withPerm(ADMIN)
+    {
+        _validateSchedule(_beneficiary, _numberOfTokens, _duration, _frequency, _startTime);
+        require(_index < dataMap[_beneficiary].schedules.length, "Schedule not found");
+        Schedule storage schedule = dataMap[_beneficiary].schedules[_index];
+        /*solium-disable-next-line security/no-block-members*/
+        require(now < schedule.startTime, "Schedule started");
+        if (_numberOfTokens <= schedule.lockedTokens) {
+            unassignedTokens = unassignedTokens.add(schedule.lockedTokens - _numberOfTokens);
+        } else {
+            require((_numberOfTokens - schedule.lockedTokens) <= unassignedTokens, "Not enough tokens");
+            unassignedTokens = unassignedTokens.sub(_numberOfTokens - schedule.lockedTokens);
+        }
+        schedule.numberOfTokens = _numberOfTokens;
+        schedule.lockedTokens = _numberOfTokens;
+        schedule.duration = _duration;
+        schedule.frequency = _frequency;
+        schedule.startTime = _startTime;
+        schedule.nextTime = _startTime.add(schedule.frequency);
+        emit EditSchedule(_beneficiary, _index, _numberOfTokens, _duration, _frequency, _startTime, now);
+    }
+
+    /**
+     * @notice Revokes beneficiary's schedule
+     * @param _beneficiary beneficiary's address
+     * @param _index index of the schedule
+     */
+    function revokeSchedule(address _beneficiary, uint256 _index) external withPerm(ADMIN) {
+        require(_beneficiary != address(0), "Invalid address");
+        require(_index < dataMap[_beneficiary].schedules.length, "Schedule not found");
+        Schedule[] storage schedules = dataMap[_beneficiary].schedules;
+        unassignedTokens = unassignedTokens.add(schedules[_index].lockedTokens);
+        schedules[_index] = schedules[schedules.length - 1];
+        schedules.length--;
+        if (schedules.length == 0) {
+            _revokeSchedules(_beneficiary);
```

Comment:
If issuer revoking the schedule then it also transfers all the unlocked tokens as well. Suppose you as an Issuer added me in a vesting schedule with a one year cliff and I left your company and did not pull my tokens yet but you revoke my schedule. then there is no way I can get my tokens. So I prefer whenever the issuer revoking the schedule function will check if there are any available tokens to send, it first transfers those tokens and then revokes the  schedule.

Reply:
It's possible to pull available tokens by a beneficiary even if schedule is revoked.
It was one of the reason for complex structures: 
    struct Data {
        uint256 index;
        Schedule[] schedules;
        uint256 availableTokens;
        uint256 claimedTokens;
    }
Anyway I will simplify sctructure and add pushing tokens in revoking time.